### DISCUSSION TITLE
What's the intuition behind using blip with BLIT_2BPP?

### DISCUSSION THREAD

**Discussion by tupini07** [aduros/wasm4#422] [Questions] [answered]
Hi all!

I was trying to better understand how `blip` works when using `BLIT_2BPP`. To do this I was trying to manually come up with an array of numbers to draw a single color from my pallette, but didn't manage to do it. I was trying something similar to:

```rust
*PALETTE = [0x000000, 0xFF0000, 0x00FF00, 0x0000FF];
*DRAW_COLORS = 0x0123;

// then pretty much try to understand if there's a 'magic number' I can use to index each 
// of the colors in `DRAW_COLORS`

let solid_red: [u8 ; 64] = [ /* is there a specific number I can repeat here to draw all pixels red? */ ]
blit(&solid_red, 10, 10, 8, 8, BLIT_2BPP);
```

Could someone offer me some intuition/example as to how `blip(.. BLIT_2BPP)` works under the hood? I took a look at [png2src](https://github.com/aduros/wasm4/blob/main/cli/lib/png2src.js) but it left me with more questions than answers 😅

**Comment by aduros** [comment, marked as answer]
Hello!

Consider the 8 bits in a byte: `a a b b c c d d`. They can divided into 2 bit chunks.

`aa` is the first pixel (representing a color value between 0 and 3).
`bb` is the second...
`cc` is the third...
`dd` is the fourth...

We use 2 bits per pixel color (this is why it's called 2BPP!), or in other words, each byte contains 4 pixel colors.

So in your example, to draw an 8x8 image, you only need an array of 16 bytes, not 64.

`DRAW_COLORS` is a level of indirection that allows you to change the colors that get drawn without modifying the entire image data. So for each color in the image, it looks up the equivalent entry in `DRAW_COLORS`, then *that* color gets put into the framebuffer. And from there, the color that gets shown on the screen is based on the `PALETTE` RGB values :smiley:

**Reply by tupini07** [reply]
oh, got it! I appreciate the explanation. I noticed I was doing a couple of things wrong:

- For some reason I was assuming `DRAW_COLORS` had its most significant bit to the left, and that `0` was the first color of the palette instead of _transparent_
- I didn't stop and really think about the different 8bit values and what they meant, so I was assuming each stood for a single pixel, which of course doesn't make sense now that I re-read the doc.

Anyway, I was able to do what I originally set out to (draw blocks of one solid color) with this code:

```rust
unsafe {
    *PALETTE = [0x40f0f3, 0xe5b083, 0x426e5d, 0x20283d];
    *DRAW_COLORS = 0x4321;
}

fn draw_block(color: u8, x: i32, y: i32) {
    let colors = [color | (color << 2) | (color << 4) | (color << 6); 16];
    blit(&colors, x, y, 8, 8, BLIT_2BPP);
}

draw_block(1, 0, 0); // this is just to provide some contrast for color 0
draw_block(0, 1, 0);

draw_block(1, 9, 0);

draw_block(2, 18, 0);

draw_block(3, 27, 0);
```

which renders as:

![image](https://user-images.githubusercontent.com/3422347/159067488-e53009b0-f3f3-4980-9393-7f24ff6d0c66.png)


Thanks!
